The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16170068067/job/45641314936) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/45641314936) [(plain)](https://github.com/rust-lang/rust/commit/e6e3a415a4a3447915299f17954c80c56925a654/checks/45641314936/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] build_script_build test:false 0.344
error[E0603]: tuple struct constructor `Empty` is private
  --> library/core/src/iter/sources.rs:13:23
   |
13 | pub use self::empty::{Empty, empty};
   |                       ^^^^^ private tuple struct constructor
   |
  ::: library/core/src/iter/sources/empty.rs:30:21
   |
30 | pub struct Empty<T>(marker::PhantomData<fn() -> T>);
   |                     ------------------------------ a constructor is private if any of the fields is private
   |
note: the tuple struct constructor `Empty` is defined here
  --> library/core/src/iter/sources/empty.rs:30:1
   |
30 | pub struct Empty<T>(marker::PhantomData<fn() -> T>);
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: consider making the field publicly accessible
  --> library/core/src/iter/sources/empty.rs:30:21
   |
30 | pub struct Empty<T>(pub marker::PhantomData<fn() -> T>);
   |                     +++

error[E0603]: tuple struct constructor `FromFn` is private
    --> library/core/src/fmt/mod.rs:43:26
     |
43   | pub use self::builders::{FromFn, from_fn};
     |                          ^^^^^^ private tuple struct constructor
     |
    ::: library/core/src/fmt/builders.rs:1240:22
     |
1240 | pub struct FromFn<F>(F)
     |                      - a constructor is private if any of the fields is private
     |
note: the tuple struct constructor `FromFn` is defined here
    --> library/core/src/fmt/builders.rs:1240:1
     |
1240 | / pub struct FromFn<F>(F)
1241 | | where
1242 | |     F: Fn(&mut fmt::Formatter<'_>) -> fmt::Result;
     | |__________________________________________________^
help: consider making the field publicly accessible
    -->  library/core/src/fmt/builders.rs:1240:22
     |
1240 | pub struct FromFn<F>(pub F)
     |                      +++

error[E0603]: module import `range` is private
   --> library/core/src/slice/mod.rs:52:17
    |
---

error[E0603]: module `empty` is private
   --> library/core/src/iter/mod.rs:426:32
    |
426 | pub use self::sources::{Empty, empty};
    |                                ^^^^^ private module
    |
note: the module `empty` is defined here
   --> library/core/src/iter/sources.rs:1:1
    |
---
    |                  ^^^^^^^ private tuple struct constructor
    |
   ::: library/core/src/num/nonzero.rs:127:42
    |
127 | pub struct NonZero<T: ZeroablePrimitive>(T::NonZeroInner);
    |                                          --------------- a constructor is private if any of the fields is private
    |
note: the tuple struct constructor `NonZero` is defined here
   --> library/core/src/num/nonzero.rs:127:1
    |
127 | pub struct NonZero<T: ZeroablePrimitive>(T::NonZeroInner);
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: consider making the field publicly accessible
   --> library/core/src/num/nonzero.rs:127:42
    |
127 | pub struct NonZero<T: ZeroablePrimitive>(pub T::NonZeroInner);
    |                                          +++

error[E0603]: tuple struct constructor `Alignment` is private
   --> library/core/src/ptr/mod.rs:408:20
    |
---

error[E0603]: tuple struct constructor `PhantomContravariant` is private
   --> library/core/src/marker.rs:13:5
    |
13  |     PhantomContravariant, PhantomContravariantLifetime, PhantomCovariant, PhantomCovariantLifetime,
    |     ^^^^^^^^^^^^^^^^^^^^ private tuple struct constructor
    |
   ::: library/core/src/marker/variance.rs:214:40
    |
214 |     pub struct PhantomContravariant<T>(PhantomData<fn(T)>);
    |                                        ------------------ a constructor is private if any of the fields is private
    |
note: the tuple struct constructor `PhantomContravariant` is defined here
   --> library/core/src/marker/variance.rs:21:9
    |
15  | / macro_rules! phantom_type {
16  | |     ($(
17  | |         $(#[$attr:meta])*
18  | |         pub struct $name:ident <$t:ident> ($($inner:tt)*);
...   |
21  | |         pub struct $name<$t>($($inner)*) where $t: ?Sized;
    | |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...   |
94  | |     )*};
95  | | }
    | |_- in this expansion of `phantom_type!`
...
178 | / phantom_type! {
179 | |     /// Zero-sized type used to mark a type parameter as covariant.
180 | |     ///
181 | |     /// Types used as part of the return value from a function are covariant. If the type is _also_
...   |
229 | |     pub struct PhantomInvariant<T>(PhantomData<fn(T) -> T>);
230 | | }
    | |_- in this macro invocation
help: consider making the field publicly accessible
   --> library/core/src/marker/variance.rs:214:40
    |
214 |     pub struct PhantomContravariant<T>(pub PhantomData<fn(T)>);
    |                                        +++

error[E0603]: tuple struct constructor `PhantomContravariantLifetime` is private
   --> library/core/src/marker.rs:13:27
    |
13  |     PhantomContravariant, PhantomContravariantLifetime, PhantomCovariant, PhantomCovariantLifetime,
    |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ private tuple struct constructor
    |
   ::: library/core/src/marker/variance.rs:160:49
    |
160 |     pub struct PhantomContravariantLifetime<'a>(PhantomContravariant<&'a ()>);
    |                                                 ---------------------------- a constructor is private if any of the fields is private
    |
note: the tuple struct constructor `PhantomContravariantLifetime` is defined here
   --> library/core/src/marker/variance.rs:104:9
    |
97  | / macro_rules! phantom_lifetime {
98  | |     ($(
99  | |         $(#[$attr:meta])*
100 | |         pub struct $name:ident <$lt:lifetime> ($($inner:tt)*);
...   |
104 | |         pub struct $name<$lt>($($inner)*);
    | |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...   |
123 | |     )*};
124 | | }
    | |_- in this expansion of `phantom_lifetime!`
125 |
126 | / phantom_lifetime! {
127 | |     /// Zero-sized type used to mark a lifetime as covariant.
128 | |     ///
129 | |     /// Covariant lifetimes must live at least as long as declared. See [the reference][1] for more
...   |
175 | |     pub struct PhantomInvariantLifetime<'a>(PhantomInvariant<&'a ()>);
176 | | }
    | |_- in this macro invocation
help: consider making the field publicly accessible
   --> library/core/src/marker/variance.rs:160:49
    |
160 |     pub struct PhantomContravariantLifetime<'a>(pub PhantomContravariant<&'a ()>);
    |                                                 +++

error[E0603]: tuple struct constructor `PhantomCovariant` is private
   --> library/core/src/marker.rs:13:57
    |
13  |     PhantomContravariant, PhantomContravariantLifetime, PhantomCovariant, PhantomCovariantLifetime,
    |                                                         ^^^^^^^^^^^^^^^^ private tuple struct constructor
    |
   ::: library/core/src/marker/variance.rs:196:36
    |
196 |     pub struct PhantomCovariant<T>(PhantomData<fn() -> T>);
    |                                    ---------------------- a constructor is private if any of the fields is private
    |
note: the tuple struct constructor `PhantomCovariant` is defined here
   --> library/core/src/marker/variance.rs:21:9
    |
15  | / macro_rules! phantom_type {
16  | |     ($(
17  | |         $(#[$attr:meta])*
18  | |         pub struct $name:ident <$t:ident> ($($inner:tt)*);
...   |
21  | |         pub struct $name<$t>($($inner)*) where $t: ?Sized;
    | |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...   |
94  | |     )*};
95  | | }
    | |_- in this expansion of `phantom_type!`
...
178 | / phantom_type! {
179 | |     /// Zero-sized type used to mark a type parameter as covariant.
180 | |     ///
181 | |     /// Types used as part of the return value from a function are covariant. If the type is _also_
...   |
229 | |     pub struct PhantomInvariant<T>(PhantomData<fn(T) -> T>);
230 | | }
    | |_- in this macro invocation
help: consider making the field publicly accessible
   --> library/core/src/marker/variance.rs:196:36
    |
196 |     pub struct PhantomCovariant<T>(pub PhantomData<fn() -> T>);
    |                                    +++

error[E0603]: tuple struct constructor `PhantomCovariantLifetime` is private
   --> library/core/src/marker.rs:13:75
    |
13  |     PhantomContravariant, PhantomContravariantLifetime, PhantomCovariant, PhantomCovariantLifetime,
    |                                                                           ^^^^^^^^^^^^^^^^^^^^^^^^ private tuple struct constructor
    |
   ::: library/core/src/marker/variance.rs:143:45
    |
143 |     pub struct PhantomCovariantLifetime<'a>(PhantomCovariant<&'a ()>);
    |                                             ------------------------ a constructor is private if any of the fields is private
    |
note: the tuple struct constructor `PhantomCovariantLifetime` is defined here
   --> library/core/src/marker/variance.rs:104:9
    |
97  | / macro_rules! phantom_lifetime {
98  | |     ($(
99  | |         $(#[$attr:meta])*
100 | |         pub struct $name:ident <$lt:lifetime> ($($inner:tt)*);
...   |
104 | |         pub struct $name<$lt>($($inner)*);
    | |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...   |
123 | |     )*};
124 | | }
    | |_- in this expansion of `phantom_lifetime!`
125 |
126 | / phantom_lifetime! {
127 | |     /// Zero-sized type used to mark a lifetime as covariant.
128 | |     ///
129 | |     /// Covariant lifetimes must live at least as long as declared. See [the reference][1] for more
...   |
175 | |     pub struct PhantomInvariantLifetime<'a>(PhantomInvariant<&'a ()>);
176 | | }
    | |_- in this macro invocation
help: consider making the field publicly accessible
   --> library/core/src/marker/variance.rs:143:45
    |
143 |     pub struct PhantomCovariantLifetime<'a>(pub PhantomCovariant<&'a ()>);
    |                                             +++

error[E0603]: tuple struct constructor `PhantomInvariant` is private
   --> library/core/src/marker.rs:14:5
    |
14  |     PhantomInvariant, PhantomInvariantLifetime, Variance, variance,
    |     ^^^^^^^^^^^^^^^^ private tuple struct constructor
    |
   ::: library/core/src/marker/variance.rs:229:36
    |
229 |     pub struct PhantomInvariant<T>(PhantomData<fn(T) -> T>);
    |                                    ----------------------- a constructor is private if any of the fields is private
    |
note: the tuple struct constructor `PhantomInvariant` is defined here
   --> library/core/src/marker/variance.rs:21:9
    |
15  | / macro_rules! phantom_type {
16  | |     ($(
17  | |         $(#[$attr:meta])*
18  | |         pub struct $name:ident <$t:ident> ($($inner:tt)*);
...   |
21  | |         pub struct $name<$t>($($inner)*) where $t: ?Sized;
    | |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...   |
94  | |     )*};
95  | | }
    | |_- in this expansion of `phantom_type!`
...
178 | / phantom_type! {
179 | |     /// Zero-sized type used to mark a type parameter as covariant.
180 | |     ///
181 | |     /// Types used as part of the return value from a function are covariant. If the type is _also_
...   |
229 | |     pub struct PhantomInvariant<T>(PhantomData<fn(T) -> T>);
230 | | }
    | |_- in this macro invocation
help: consider making the field publicly accessible
   --> library/core/src/marker/variance.rs:229:36
    |
229 |     pub struct PhantomInvariant<T>(pub PhantomData<fn(T) -> T>);
    |                                    +++

error[E0603]: tuple struct constructor `PhantomInvariantLifetime` is private
   --> library/core/src/marker.rs:14:23
    |
14  |     PhantomInvariant, PhantomInvariantLifetime, Variance, variance,
    |                       ^^^^^^^^^^^^^^^^^^^^^^^^ private tuple struct constructor
    |
   ::: library/core/src/marker/variance.rs:175:45
    |
175 |     pub struct PhantomInvariantLifetime<'a>(PhantomInvariant<&'a ()>);
    |                                             ------------------------ a constructor is private if any of the fields is private
    |
note: the tuple struct constructor `PhantomInvariantLifetime` is defined here
   --> library/core/src/marker/variance.rs:104:9
    |
97  | / macro_rules! phantom_lifetime {
98  | |     ($(
99  | |         $(#[$attr:meta])*
100 | |         pub struct $name:ident <$lt:lifetime> ($($inner:tt)*);
...   |
104 | |         pub struct $name<$lt>($($inner)*);
    | |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...   |
123 | |     )*};
124 | | }
    | |_- in this expansion of `phantom_lifetime!`
125 |
126 | / phantom_lifetime! {
127 | |     /// Zero-sized type used to mark a lifetime as covariant.
128 | |     ///
129 | |     /// Covariant lifetimes must live at least as long as declared. See [the reference][1] for more
...   |
175 | |     pub struct PhantomInvariantLifetime<'a>(PhantomInvariant<&'a ()>);
176 | | }
    | |_- in this macro invocation
help: consider making the field publicly accessible
   --> library/core/src/marker/variance.rs:175:45
    |
175 |     pub struct PhantomInvariantLifetime<'a>(pub PhantomInvariant<&'a ()>);
    |                                             +++

error[E0603]: tuple struct constructor `FromBytesUntilNulError` is private
   --> library/core/src/ffi/mod.rs:17:22
    |
17  | pub use self::c_str::FromBytesUntilNulError;
    |                      ^^^^^^^^^^^^^^^^^^^^^^ private tuple struct constructor
    |
   ::: library/core/src/ffi/c_str.rs:156:35
    |
156 | pub struct FromBytesUntilNulError(());
    |                                   -- a constructor is private if any of the fields is private
    |
note: the tuple struct constructor `FromBytesUntilNulError` is defined here
   --> library/core/src/ffi/c_str.rs:156:1
    |
156 | pub struct FromBytesUntilNulError(());
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: consider making the field publicly accessible
   --> library/core/src/ffi/c_str.rs:156:35
    |
156 | pub struct FromBytesUntilNulError(pub ());
    |                                   +++

error[E0603]: module `chain` is private
   --> library/core/src/iter/mod.rs:406:25
    |
---

error[E0603]: module `from_fn` is private
   --> library/core/src/iter/mod.rs:434:33
    |
434 | pub use self::sources::{FromFn, from_fn};
    |                                 ^^^^^^^ private module
    |
note: the module `from_fn` is defined here
   --> library/core/src/iter/sources.rs:3:1
    |
3   | mod from_fn;
    | ^^^^^^^^^^^^

error[E0603]: module `once` is private
   --> library/core/src/iter/mod.rs:436:31
    |
436 | pub use self::sources::{Once, once};
    |                               ^^^^ private module
    |
note: the module `once` is defined here
   --> library/core/src/iter/sources.rs:5:1
    |
5   | mod once;
    | ^^^^^^^^^

error[E0603]: module `once_with` is private
   --> library/core/src/iter/mod.rs:438:35
    |
438 | pub use self::sources::{OnceWith, once_with};
    |                                   ^^^^^^^^^ private module
    |
note: the module `once_with` is defined here
   --> library/core/src/iter/sources.rs:6:1
    |
---

error[E0603]: module `repeat_with` is private
   --> library/core/src/iter/mod.rs:444:37
    |
444 | pub use self::sources::{RepeatWith, repeat_with};
    |                                     ^^^^^^^^^^^ private module
    |
note: the module `repeat_with` is defined here
   --> library/core/src/iter/sources.rs:9:1
    |
9   | mod repeat_with;
    | ^^^^^^^^^^^^^^^^

error[E0603]: module `successors` is private
   --> library/core/src/iter/mod.rs:446:37
    |
446 | pub use self::sources::{Successors, successors};
    |                                     ^^^^^^^^^^ private module
    |
note: the module `successors` is defined here
   --> library/core/src/iter/sources.rs:10:1
    |
---
   |                                ^^^^^^^^^^^^^ private tuple struct constructor
   |
  ::: library/core/src/iter/sources/from_coroutine.rs:39:29
   |
39 | pub struct FromCoroutine<G>(G);
   |                             - a constructor is private if any of the fields is private
   |
note: the tuple struct constructor `FromCoroutine` is defined here
  --> library/core/src/iter/sources/from_coroutine.rs:39:1
   |
39 | pub struct FromCoroutine<G>(G);
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: consider making the field publicly accessible
  --> library/core/src/iter/sources/from_coroutine.rs:39:29
   |
39 | pub struct FromCoroutine<G>(pub G);
   |                             +++

error[E0603]: tuple struct constructor `FromFn` is private
  --> library/core/src/iter/sources.rs:21:25
   |
21 | pub use self::from_fn::{FromFn, from_fn};
   |                         ^^^^^^ private tuple struct constructor
   |
  ::: library/core/src/iter/sources/from_fn.rs:60:22
   |
60 | pub struct FromFn<F>(F);
   |                      - a constructor is private if any of the fields is private
   |
note: the tuple struct constructor `FromFn` is defined here
  --> library/core/src/iter/sources/from_fn.rs:60:1
   |
60 | pub struct FromFn<F>(F);
   | ^^^^^^^^^^^^^^^^^^^^^^^^
help: consider making the field publicly accessible
  --> library/core/src/iter/sources/from_fn.rs:60:22
   |
60 | pub struct FromFn<F>(pub F);
   |                      +++

error[E0603]: tuple struct constructor `AddrParseError` is private
   --> library/core/src/net/mod.rs:17:23
    |
---
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: consider making the field publicly accessible
   --> library/core/src/net/parser.rs:496:27
    |
496 | pub struct AddrParseError(pub AddrKind);
    |                           +++

error[E0603]: tuple struct constructor `IterRange` is private
  --> library/core/src/range.rs:29:16
   |
29 | pub use iter::{IterRange, IterRangeFrom, IterRangeInclusive};
   |                ^^^^^^^^^ private tuple struct constructor
   |
  ::: library/core/src/range/iter.rs:10:25
   |
10 | pub struct IterRange<A>(legacy::Range<A>);
   |                         ---------------- a constructor is private if any of the fields is private
   |
note: the tuple struct constructor `IterRange` is defined here
  --> library/core/src/range/iter.rs:10:1
   |
10 | pub struct IterRange<A>(legacy::Range<A>);
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: consider making the field publicly accessible
  --> library/core/src/range/iter.rs:10:25
   |
10 | pub struct IterRange<A>(pub legacy::Range<A>);
   |                         +++

error[E0603]: tuple struct constructor `IterRangeFrom` is private
   --> library/core/src/range.rs:29:27
    |
29  | pub use iter::{IterRange, IterRangeFrom, IterRangeInclusive};
    |                           ^^^^^^^^^^^^^ private tuple struct constructor
    |
   ::: library/core/src/range/iter.rs:296:29
    |
296 | pub struct IterRangeFrom<A>(legacy::RangeFrom<A>);
    |                             -------------------- a constructor is private if any of the fields is private
    |
note: the tuple struct constructor `IterRangeFrom` is defined here
   --> library/core/src/range/iter.rs:296:1
    |
296 | pub struct IterRangeFrom<A>(legacy::RangeFrom<A>);
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: consider making the field publicly accessible
   --> library/core/src/range/iter.rs:296:29
    |
296 | pub struct IterRangeFrom<A>(pub legacy::RangeFrom<A>);
    |                             +++

error[E0603]: tuple struct constructor `IterRangeInclusive` is private
   --> library/core/src/range.rs:29:42
    |
29  | pub use iter::{IterRange, IterRangeFrom, IterRangeInclusive};
    |                                          ^^^^^^^^^^^^^^^^^^ private tuple struct constructor
    |
   ::: library/core/src/range/iter.rs:156:34
    |
156 | pub struct IterRangeInclusive<A>(legacy::RangeInclusive<A>);
    |                                  ------------------------- a constructor is private if any of the fields is private
    |
note: the tuple struct constructor `IterRangeInclusive` is defined here
   --> library/core/src/range/iter.rs:156:1
    |
156 | pub struct IterRangeInclusive<A>(legacy::RangeInclusive<A>);
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: consider making the field publicly accessible
   --> library/core/src/range/iter.rs:156:34
    |
156 | pub struct IterRangeInclusive<A>(pub legacy::RangeInclusive<A>);
    |                                  +++

error[E0603]: tuple struct constructor `SipHasher` is private
  --> library/core/src/hash/mod.rs:88:20
   |
88 | pub use self::sip::SipHasher;
   |                    ^^^^^^^^^ private tuple struct constructor
   |
  ::: library/core/src/hash/sip.rs:47:22
   |
47 | pub struct SipHasher(SipHasher24);
   |                      ----------- a constructor is private if any of the fields is private
   |
note: the tuple struct constructor `SipHasher` is defined here
  --> library/core/src/hash/sip.rs:47:1
   |
47 | pub struct SipHasher(SipHasher24);
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: consider making the field publicly accessible
  --> library/core/src/hash/sip.rs:47:22
   |
47 | pub struct SipHasher(pub SipHasher24);
   |                      +++

error[E0603]: tuple struct constructor `EscapeDefault` is private
  --> library/core/src/slice/ascii.rs:3:18
   |
3  | use core::ascii::EscapeDefault;
   |                  ^^^^^^^^^^^^^ private tuple struct constructor
   |
  ::: library/core/src/ascii.rs:29:26
   |
29 | pub struct EscapeDefault(EscapeIterInner<4, AlwaysEscaped>);
   |                          --------------------------------- a constructor is private if any of the fields is private
   |
note: the tuple struct constructor `EscapeDefault` is defined here
  --> library/core/src/ascii.rs:29:1
   |
29 | pub struct EscapeDefault(EscapeIterInner<4, AlwaysEscaped>);
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: consider making the field publicly accessible
  --> library/core/src/ascii.rs:29:26
   |
29 | pub struct EscapeDefault(pub EscapeIterInner<4, AlwaysEscaped>);
   |                          +++

error[E0603]: tuple struct constructor `Nanoseconds` is private
   --> library/core/src/time.rs:24:30
    |
24  | use crate::num::niche_types::Nanoseconds;
    |                              ^^^^^^^^^^^ private tuple struct constructor
    |
   ::: library/core/src/num/niche_types.rs:22:27
    |
22  |         $vis struct $name($int);
    |                           ---- a constructor is private if any of the fields is private
    |
note: the tuple struct constructor `Nanoseconds` is defined here
   --> library/core/src/num/niche_types.rs:22:9
    |
12  | / macro_rules! define_valid_range_type {
13  | |     ($(
14  | |         $(#[$m:meta])*
15  | |         $vis:vis struct $name:ident($int:ident as $uint:ident in $low:literal..=$high:literal);
...   |
22  | |         $vis struct $name($int);
    | |         ^^^^^^^^^^^^^^^^^^^^^^^^
...   |
103 | |     )+};
104 | | }
    | |_- in this expansion of `define_valid_range_type!`
105 |
106 | / define_valid_range_type! {
107 | |     pub struct Nanoseconds(u32 as u32 in 0..=999_999_999);
108 | | }
    | |_- in this macro invocation
help: consider making the field publicly accessible
   --> library/core/src/num/niche_types.rs:22:27
    |
22  |         $vis struct $name(pub $int);
    |                           +++

error[E0603]: tuple struct constructor `Ready` is private
  --> library/core/src/future/mod.rs:32:17
   |
32 | pub use ready::{Ready, ready};
   |                 ^^^^^ private tuple struct constructor
   |
  ::: library/core/src/future/ready.rs:12:21
   |
12 | pub struct Ready<T>(Option<T>);
   |                     --------- a constructor is private if any of the fields is private
   |
note: the tuple struct constructor `Ready` is defined here
  --> library/core/src/future/ready.rs:12:1
   |
12 | pub struct Ready<T>(Option<T>);
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: consider making the field publicly accessible
  --> library/core/src/future/ready.rs:12:21
   |
12 | pub struct Ready<T>(pub Option<T>);
   |                     +++

error[E0603]: tuple struct constructor `i8x16` is private
   --> library/core/src/../../stdarch/crates/core_arch/src/x86/avx512bitalg.rs:10:29
    |
10  | use crate::core_arch::simd::i8x16;
    |                             ^^^^^ private tuple struct constructor
    |
   ::: library/core/src/../../stdarch/crates/core_arch/src/simd.rs:9:31
    |
9   |         pub(crate) struct $id([$elem_type; $len]);
    |                               ------------------ a constructor is private if any of the fields is private
    |
note: the tuple struct constructor `i8x16` is defined here
   --> library/core/src/../../stdarch/crates/core_arch/src/simd.rs:9:9
    |
5   | / macro_rules! simd_ty {
6   | |     ($id:ident [$elem_type:ty ; $len:literal]: $($param_name:ident),*) => {
7   | |         #[repr(simd)]
8   | |         #[derive(Copy, Clone)]
9   | |         pub(crate) struct $id([$elem_type; $len]);
    | |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...   |
70  | | }
    | |_- in this expansion of `simd_ty!`
...
212 | / simd_ty!(
213 | |     i8x16[i8;16]:
214 | |     x0,
215 | |     x1,
...   |
229 | |     x15
230 | | );
    | |_- in this macro invocation
help: consider making the field publicly accessible
   --> library/core/src/../../stdarch/crates/core_arch/src/simd.rs:9:31
    |
9   |         pub(crate) struct $id(pub [$elem_type; $len]);
    |                               +++

error[E0603]: tuple struct constructor `i8x32` is private
   --> library/core/src/../../stdarch/crates/core_arch/src/x86/avx512bitalg.rs:11:29
    |
11  | use crate::core_arch::simd::i8x32;
    |                             ^^^^^ private tuple struct constructor
    |
   ::: library/core/src/../../stdarch/crates/core_arch/src/simd.rs:9:31
    |
9   |         pub(crate) struct $id([$elem_type; $len]);
    |                               ------------------ a constructor is private if any of the fields is private
    |
note: the tuple struct constructor `i8x32` is defined here
   --> library/core/src/../../stdarch/crates/core_arch/src/simd.rs:9:9
    |
5   | / macro_rules! simd_ty {
6   | |     ($id:ident [$elem_type:ty ; $len:literal]: $($param_name:ident),*) => {
7   | |         #[repr(simd)]
8   | |         #[derive(Copy, Clone)]
9   | |         pub(crate) struct $id([$elem_type; $len]);
    | |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...   |
70  | | }
    | |_- in this expansion of `simd_ty!`
...
363 | / simd_ty!(
364 | |     i8x32[i8;32]:
365 | |     x0,
366 | |     x1,
...   |
396 | |     x31
397 | | );
    | |_- in this macro invocation
help: consider making the field publicly accessible
   --> library/core/src/../../stdarch/crates/core_arch/src/simd.rs:9:31
    |
9   |         pub(crate) struct $id(pub [$elem_type; $len]);
    |                               +++

error[E0603]: tuple struct constructor `i8x64` is private
   --> library/core/src/../../stdarch/crates/core_arch/src/x86/avx512bitalg.rs:12:29
    |
12  | use crate::core_arch::simd::i8x64;
    |                             ^^^^^ private tuple struct constructor
    |
   ::: library/core/src/../../stdarch/crates/core_arch/src/simd.rs:9:31
    |
9   |         pub(crate) struct $id([$elem_type; $len]);
    |                               ------------------ a constructor is private if any of the fields is private
    |
note: the tuple struct constructor `i8x64` is defined here
   --> library/core/src/../../stdarch/crates/core_arch/src/simd.rs:9:9
    |
5   | / macro_rules! simd_ty {
6   | |     ($id:ident [$elem_type:ty ; $len:literal]: $($param_name:ident),*) => {
7   | |         #[repr(simd)]
8   | |         #[derive(Copy, Clone)]
9   | |         pub(crate) struct $id([$elem_type; $len]);
    | |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...   |
70  | | }
    | |_- in this expansion of `simd_ty!`
...
530 | / simd_ty!(
531 | |     i8x64[i8;64]:
532 | |     x0,
533 | |     x1,
...   |
595 | |     x63
596 | | );
    | |_- in this macro invocation
help: consider making the field publicly accessible
   --> library/core/src/../../stdarch/crates/core_arch/src/simd.rs:9:31
    |
9   |         pub(crate) struct $id(pub [$elem_type; $len]);
    |                               +++

error[E0603]: tuple struct constructor `i16x8` is private
   --> library/core/src/../../stdarch/crates/core_arch/src/x86/avx512bitalg.rs:13:29
    |
13  | use crate::core_arch::simd::i16x8;
    |                             ^^^^^ private tuple struct constructor
    |
   ::: library/core/src/../../stdarch/crates/core_arch/src/simd.rs:9:31
    |
9   |         pub(crate) struct $id([$elem_type; $len]);
    |                               ------------------ a constructor is private if any of the fields is private
    |
note: the tuple struct constructor `i16x8` is defined here
   --> library/core/src/../../stdarch/crates/core_arch/src/simd.rs:9:9
    |
5   | / macro_rules! simd_ty {
6   | |     ($id:ident [$elem_type:ty ; $len:literal]: $($param_name:ident),*) => {
7   | |         #[repr(simd)]
8   | |         #[derive(Copy, Clone)]
9   | |         pub(crate) struct $id([$elem_type; $len]);
    | |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...   |
70  | | }
    | |_- in this expansion of `simd_ty!`
...
231 | / simd_ty!(
232 | |     i16x8[i16;8]:
233 | |     x0,
234 | |     x1,
...   |
240 | |     x7
241 | | );
    | |_- in this macro invocation
help: consider making the field publicly accessible
   --> library/core/src/../../stdarch/crates/core_arch/src/simd.rs:9:31
    |
9   |         pub(crate) struct $id(pub [$elem_type; $len]);
    |                               +++

error[E0603]: tuple struct constructor `i16x16` is private
   --> library/core/src/../../stdarch/crates/core_arch/src/x86/avx512bitalg.rs:14:29
    |
14  | use crate::core_arch::simd::i16x16;
    |                             ^^^^^^ private tuple struct constructor
    |
   ::: library/core/src/../../stdarch/crates/core_arch/src/simd.rs:9:31
    |
9   |         pub(crate) struct $id([$elem_type; $len]);
    |                               ------------------ a constructor is private if any of the fields is private
    |
note: the tuple struct constructor `i16x16` is defined here
   --> library/core/src/../../stdarch/crates/core_arch/src/simd.rs:9:9
    |
5   | / macro_rules! simd_ty {
6   | |     ($id:ident [$elem_type:ty ; $len:literal]: $($param_name:ident),*) => {
7   | |         #[repr(simd)]
8   | |         #[derive(Copy, Clone)]
9   | |         pub(crate) struct $id([$elem_type; $len]);
    | |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...   |
70  | | }
    | |_- in this expansion of `simd_ty!`
...
398 | / simd_ty!(
399 | |     i16x16[i16;16]:
400 | |     x0,
401 | |     x1,
...   |
415 | |     x15
416 | | );
    | |_- in this macro invocation
help: consider making the field publicly accessible
   --> library/core/src/../../stdarch/crates/core_arch/src/simd.rs:9:31
    |
9   |         pub(crate) struct $id(pub [$elem_type; $len]);
    |                               +++

error[E0603]: tuple struct constructor `i16x32` is private
   --> library/core/src/../../stdarch/crates/core_arch/src/x86/avx512bitalg.rs:15:29
    |
15  | use crate::core_arch::simd::i16x32;
    |                             ^^^^^^ private tuple struct constructor
    |
   ::: library/core/src/../../stdarch/crates/core_arch/src/simd.rs:9:31
    |
9   |         pub(crate) struct $id([$elem_type; $len]);
    |                               ------------------ a constructor is private if any of the fields is private
    |
note: the tuple struct constructor `i16x32` is defined here
   --> library/core/src/../../stdarch/crates/core_arch/src/simd.rs:9:9
    |
5   | / macro_rules! simd_ty {
6   | |     ($id:ident [$elem_type:ty ; $len:literal]: $($param_name:ident),*) => {
7   | |         #[repr(simd)]
8   | |         #[derive(Copy, Clone)]
9   | |         pub(crate) struct $id([$elem_type; $len]);
    | |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...   |
70  | | }
    | |_- in this expansion of `simd_ty!`
...
666 | / simd_ty!(
667 | |     i16x32[i16;32]:
668 | |     x0,
669 | |     x1,
...   |
699 | |     x31
700 | | );
    | |_- in this macro invocation
help: consider making the field publicly accessible
   --> library/core/src/../../stdarch/crates/core_arch/src/simd.rs:9:31
    |
9   |         pub(crate) struct $id(pub [$elem_type; $len]);
    |                               +++

error[E0603]: tuple struct constructor `i8x16` is private
   --> library/core/src/../../stdarch/crates/core_arch/src/x86/gfni.rs:10:29
    |
10  | use crate::core_arch::simd::i8x16;
    |                             ^^^^^ private tuple struct constructor
    |
   ::: library/core/src/../../stdarch/crates/core_arch/src/simd.rs:9:31
    |
9   |         pub(crate) struct $id([$elem_type; $len]);
    |                               ------------------ a constructor is private if any of the fields is private
    |
note: the tuple struct constructor `i8x16` is defined here
   --> library/core/src/../../stdarch/crates/core_arch/src/simd.rs:9:9
    |
5   | / macro_rules! simd_ty {
6   | |     ($id:ident [$elem_type:ty ; $len:literal]: $($param_name:ident),*) => {
7   | |         #[repr(simd)]
8   | |         #[derive(Copy, Clone)]
9   | |         pub(crate) struct $id([$elem_type; $len]);
    | |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...   |
70  | | }
    | |_- in this expansion of `simd_ty!`
...
212 | / simd_ty!(
213 | |     i8x16[i8;16]:
214 | |     x0,
215 | |     x1,
...   |
229 | |     x15
230 | | );
    | |_- in this macro invocation
help: consider making the field publicly accessible
   --> library/core/src/../../stdarch/crates/core_arch/src/simd.rs:9:31
    |
9   |         pub(crate) struct $id(pub [$elem_type; $len]);
    |                               +++

error[E0603]: tuple struct constructor `i8x32` is private
   --> library/core/src/../../stdarch/crates/core_arch/src/x86/gfni.rs:11:29
    |
11  | use crate::core_arch::simd::i8x32;
    |                             ^^^^^ private tuple struct constructor
    |
   ::: library/core/src/../../stdarch/crates/core_arch/src/simd.rs:9:31
    |
9   |         pub(crate) struct $id([$elem_type; $len]);
    |                               ------------------ a constructor is private if any of the fields is private
    |
note: the tuple struct constructor `i8x32` is defined here
   --> library/core/src/../../stdarch/crates/core_arch/src/simd.rs:9:9
    |
5   | / macro_rules! simd_ty {
6   | |     ($id:ident [$elem_type:ty ; $len:literal]: $($param_name:ident),*) => {
7   | |         #[repr(simd)]
8   | |         #[derive(Copy, Clone)]
9   | |         pub(crate) struct $id([$elem_type; $len]);
    | |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...   |
70  | | }
    | |_- in this expansion of `simd_ty!`
...
363 | / simd_ty!(
364 | |     i8x32[i8;32]:
365 | |     x0,
366 | |     x1,
...   |
396 | |     x31
397 | | );
    | |_- in this macro invocation
help: consider making the field publicly accessible
   --> library/core/src/../../stdarch/crates/core_arch/src/simd.rs:9:31
    |
9   |         pub(crate) struct $id(pub [$elem_type; $len]);
    |                               +++

error[E0603]: tuple struct constructor `i8x64` is private
   --> library/core/src/../../stdarch/crates/core_arch/src/x86/gfni.rs:12:29
    |
12  | use crate::core_arch::simd::i8x64;
    |                             ^^^^^ private tuple struct constructor
    |
   ::: library/core/src/../../stdarch/crates/core_arch/src/simd.rs:9:31
    |
9   |         pub(crate) struct $id([$elem_type; $len]);
    |                               ------------------ a constructor is private if any of the fields is private
    |
note: the tuple struct constructor `i8x64` is defined here
   --> library/core/src/../../stdarch/crates/core_arch/src/simd.rs:9:9
    |
5   | / macro_rules! simd_ty {
6   | |     ($id:ident [$elem_type:ty ; $len:literal]: $($param_name:ident),*) => {
7   | |         #[repr(simd)]
8   | |         #[derive(Copy, Clone)]
9   | |         pub(crate) struct $id([$elem_type; $len]);
    | |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...   |
---
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: consider making the field publicly accessible
   --> library/core/src/char/convert.rs:265:29
    |
265 | pub struct CharTryFromError(pub ());
    |                             +++

For more information about this error, try `rustc --explain E0603`.
[RUSTC-TIMING] core test:false 21.736
error: could not compile `core` (lib) due to 44 previous errors
```

</details>

--- a/compiler/rustc_resolve/src/build_reduced_graph.rs
+++ b/compiler/rustc_resolve/src/build_reduced_graph.rs
@@ -32,10 +32,9 @@ use crate::def_collector::collect_definitions;
 use crate::imports::{ImportData, ImportKind};
 use crate::macros::{MacroRulesBinding, MacroRulesScope, MacroRulesScopeRef};
 use crate::{
-    BindingKey, Determinacy, ExternPreludeEntry, Finalize, MacroData, Module, ModuleKind,
-    ModuleOrUniformRoot, NameBinding, NameBindingData, NameBindingKind, ParentScope, PathResult,
-    ResolutionError, Resolver, ResolverArenas, Segment, ToNameBinding, Used, VisResolutionError,
-    errors,
+    BindingKey, ExternPreludeEntry, Finalize, MacroData, Module, ModuleKind, ModuleOrUniformRoot,
+    NameBinding, NameBindingData, NameBindingKind, ParentScope, PathResult, ResolutionError,
+    Resolver, ResolverArenas, Segment, ToNameBinding, Used, VisResolutionError, errors,
 };
 
 type Res = def::Res<NodeId>;
@@ -620,16 +619,7 @@ impl<'a, 'ra, 'tcx> BuildReducedGraphVisitor<'a, 'ra, 'tcx> {
                 let kind = ImportKind::Single {
                     source: source.ident,
                     target: ident,
-                    source_bindings: PerNS {
-                        type_ns: Cell::new(Err(Determinacy::Undetermined)),
-                        value_ns: Cell::new(Err(Determinacy::Undetermined)),
-                        macro_ns: Cell::new(Err(Determinacy::Undetermined)),
-                    },
-                    target_bindings: PerNS {
-                        type_ns: Cell::new(None),
-                        value_ns: Cell::new(None),
-                        macro_ns: Cell::new(None),
-                    },
+                    bindings: Default::default(),
                     type_ns_only,
                     nested,
                     id,

--- a/compiler/rustc_resolve/src/ident.rs
+++ b/compiler/rustc_resolve/src/ident.rs
@@ -1090,14 +1090,13 @@ impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
             let Some(module) = single_import.imported_module.get() else {
                 return true;
             };
-            let ImportKind::Single { source, target, target_bindings, .. } = &single_import.kind
-            else {
+            let ImportKind::Single { source, target, bindings, .. } = &single_import.kind else {
                 unreachable!();
             };
             if source != target {
-                if target_bindings.iter().all(|binding| binding.get().is_none()) {
+                if bindings.iter().all(|binding| binding.get().binding().is_none()) {
                     return true;
-                } else if target_bindings[ns].get().is_none() && binding.is_some() {
+                } else if bindings[ns].get().binding().is_none() && binding.is_some() {
                     return true;
                 }
             }

--- a/compiler/rustc_resolve/src/imports.rs
+++ b/compiler/rustc_resolve/src/imports.rs
@@ -24,7 +24,6 @@ use rustc_span::{Ident, Span, Symbol, kw, sym};
 use smallvec::SmallVec;
 use tracing::debug;
 
-use crate::Determinacy::{self, *};
 use crate::Namespace::*;
 use crate::diagnostics::{DiagMode, Suggestion, import_candidates};
 use crate::errors::{
@@ -33,12 +32,29 @@ use crate::errors::{
     ConsiderAddingMacroExport, ConsiderMarkingAsPub,
 };
 use crate::{
-    AmbiguityError, AmbiguityKind, BindingKey, Finalize, ImportSuggestion, Module,
+    AmbiguityError, AmbiguityKind, BindingKey, Determinacy, Finalize, ImportSuggestion, Module,
     ModuleOrUniformRoot, NameBinding, NameBindingData, NameBindingKind, ParentScope, PathResult,
     PerNS, ResolutionError, Resolver, ScopeSet, Segment, Used, module_to_string, names_to_string,
 };
 
 type Res = def::Res<NodeId>;
+
+/// A [`NameBinding`] in the process of being resolved.
+#[derive(Clone, Copy, Default, PartialEq)]
+pub(crate) enum PendingBinding<'ra> {
+    Ready(Option<NameBinding<'ra>>),
+    #[default]
+    Pending,
+}
+
+impl<'ra> PendingBinding<'ra> {
+    pub(crate) fn binding(self) -> Option<NameBinding<'ra>> {
+        match self {
+            PendingBinding::Ready(binding) => binding,
+            PendingBinding::Pending => None,
+        }
+    }
+}
 
 /// Contains data for specific kinds of imports.
 #[derive(Clone)]
@@ -49,10 +65,8 @@ pub(crate) enum ImportKind<'ra> {
         /// `target` in `use prefix::source as target`.
         /// It will directly use `source` when the format is `use prefix::source`.
         target: Ident,
-        /// Bindings to which `source` refers to.
-        source_bindings: PerNS<Cell<Result<NameBinding<'ra>, Determinacy>>>,
-        /// Bindings introduced by `target`.
-        target_bindings: PerNS<Cell<Option<NameBinding<'ra>>>>,
+        /// Bindings introduced by the import.
+        bindings: PerNS<Cell<PendingBinding<'ra>>>,
         /// `true` for `...::{self [as target]}` imports, `false` otherwise.
         type_ns_only: bool,
         /// Did this import result from a nested import? ie. `use foo::{bar, baz};`
@@ -96,26 +110,14 @@ impl<'ra> std::fmt::Debug for ImportKind<'ra> {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         use ImportKind::*;
         match self {
-            Single {
-                source,
-                target,
-                source_bindings,
-                target_bindings,
-                type_ns_only,
-                nested,
-                id,
-            } => f
+            Single { source, target, bindings, type_ns_only, nested, id, .. } => f
                 .debug_struct("Single")
                 .field("source", source)
                 .field("target", target)
                 // Ignore the nested bindings to avoid an infinite loop while printing.
                 .field(
-                    "source_bindings",
-                    &source_bindings.clone().map(|b| b.into_inner().map(|_| format_args!(".."))),
-                )
-                .field(
-                    "target_bindings",
-                    &target_bindings.clone().map(|b| b.into_inner().map(|_| format_args!(".."))),
+                    "bindings",
+                    &bindings.clone().map(|b| b.into_inner().binding().map(|_| format_args!(".."))),
                 )
                 .field("type_ns_only", type_ns_only)
                 .field("nested", nested)
@@ -491,8 +493,9 @@ impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
     // Define a dummy resolution containing a `Res::Err` as a placeholder for a failed
     // or indeterminate resolution, also mark such failed imports as used to avoid duplicate diagnostics.
     fn import_dummy_binding(&mut self, import: Import<'ra>, is_indeterminate: bool) {
-        if let ImportKind::Single { target, ref target_bindings, .. } = import.kind {
-            if !(is_indeterminate || target_bindings.iter().all(|binding| binding.get().is_none()))
+        if let ImportKind::Single { target, ref bindings, .. } = import.kind {
+            if !(is_indeterminate
+                || bindings.iter().all(|binding| binding.get().binding().is_none()))
             {
                 return; // Has resolution, do not create the dummy binding
             }
@@ -567,10 +570,10 @@ impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
 
             glob_error |= import.is_glob();
 
-            if let ImportKind::Single { source, ref source_bindings, .. } = import.kind
+            if let ImportKind::Single { source, ref bindings, .. } = import.kind
                 && source.name == kw::SelfLower
                 // Silence `unresolved import` error if E0429 is already emitted
-                && let Err(Determined) = source_bindings.value_ns.get()
+                && let PendingBinding::Ready(None) = bindings.value_ns.get()
             {
                 continue;
             }
@@ -819,15 +822,10 @@ impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
         };
 
         import.imported_module.set(Some(module));
-        let (source, target, source_bindings, target_bindings, type_ns_only) = match import.kind {
-            ImportKind::Single {
-                source,
-                target,
-                ref source_bindings,
-                ref target_bindings,
-                type_ns_only,
-                ..
-            } => (source, target, source_bindings, target_bindings, type_ns_only),
+        let (source, target, bindings, type_ns_only) = match import.kind {
+            ImportKind::Single { source, target, ref bindings, type_ns_only, .. } => {
+                (source, target, bindings, type_ns_only)
+            }
             ImportKind::Glob { .. } => {
                 self.resolve_glob_import(import);
                 return 0;
@@ -838,21 +836,18 @@ impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
         let mut indeterminate_count = 0;
         self.per_ns(|this, ns| {
             if !type_ns_only || ns == TypeNS {
-                if let Err(Undetermined) = source_bindings[ns].get() {
-                    let binding = this.maybe_resolve_ident_in_module(
-                        module,
-                        source,
-                        ns,
-                        &import.parent_scope,
-                        Some(import),
-                    );
-                    source_bindings[ns].set(binding);
-                } else {
+                if bindings[ns].get() != PendingBinding::Pending {
                     return;
                 };
-
+                let binding_result = this.maybe_resolve_ident_in_module(
+                    module,
+                    source,
+                    ns,
+                    &import.parent_scope,
+                    Some(import),
+                );
                 let parent = import.parent_scope.module;
-                match source_bindings[ns].get() {
+                let binding = match binding_result {
                     Ok(binding) => {
                         if binding.is_assoc_item()
                             && !this.tcx.features().import_trait_associated_functions()
@@ -865,12 +860,12 @@ impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
                             )
                             .emit();
                         }
-
+                        // We need the `target`, `source` can be extracted.
                         let imported_binding = this.import(binding, import);
-                        target_bindings[ns].set(Some(imported_binding));
                         this.define(parent, target, ns, imported_binding);
+                        PendingBinding::Ready(Some(imported_binding))
                     }
-                    Err(Determined) => {
+                    Err(Determinacy::Determined) => {
                         // Don't update the resolution for underscores, because it was never added.
                         if target.name != kw::Underscore {
                             let key = BindingKey::new(target, ns);
@@ -878,9 +873,14 @@ impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
                                 resolution.single_imports.swap_remove(&import);
                             });
                         }
+                        PendingBinding::Ready(None)
                     }
-                    Err(Undetermined) => indeterminate_count += 1,
-                }
+                    Err(Determinacy::Undetermined) => {
+                        indeterminate_count += 1;
+                        PendingBinding::Pending
+                    }
+                };
+                bindings[ns].set(binding);
             }
         });
 
@@ -893,7 +893,7 @@ impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
     /// consolidate multiple unresolved import errors into a single diagnostic.
     fn finalize_import(&mut self, import: Import<'ra>) -> Option<UnresolvedImportError> {
         let ignore_binding = match &import.kind {
-            ImportKind::Single { target_bindings, .. } => target_bindings[TypeNS].get(),
+            ImportKind::Single { bindings, .. } => bindings[TypeNS].get().binding(),
             _ => None,
         };
         let ambiguity_errors_len =
@@ -1011,60 +1011,53 @@ impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
             PathResult::Indeterminate => unreachable!(),
         };
 
-        let (ident, target, source_bindings, target_bindings, type_ns_only, import_id) =
-            match import.kind {
-                ImportKind::Single {
-                    source,
-                    target,
-                    ref source_bindings,
-                    ref target_bindings,
-                    type_ns_only,
-                    id,
-                    ..
-                } => (source, target, source_bindings, target_bindings, type_ns_only, id),
-                ImportKind::Glob { is_prelude, ref max_vis, id } => {
-                    if import.module_path.len() <= 1 {
-                        // HACK(eddyb) `lint_if_path_starts_with_module` needs at least
-                        // 2 segments, so the `resolve_path` above won't trigger it.
-                        let mut full_path = import.module_path.clone();
-                        full_path.push(Segment::from_ident(Ident::dummy()));
-                        self.lint_if_path_starts_with_module(Some(finalize), &full_path, None);
-                    }
-
-                    if let ModuleOrUniformRoot::Module(module) = module
-                        && module == import.parent_scope.module
-                    {
-                        // Importing a module into itself is not allowed.
-                        return Some(UnresolvedImportError {
-                            span: import.span,
-                            label: Some(String::from("cannot glob-import a module into itself")),
-                            note: None,
-                            suggestion: None,
-                            candidates: None,
-                            segment: None,
-                            module: None,
-                        });
-                    }
-                    if !is_prelude
-                        && let Some(max_vis) = max_vis.get()
-                        && !max_vis.is_at_least(import.vis, self.tcx)
-                    {
-                        let def_id = self.local_def_id(id);
-                        self.lint_buffer.buffer_lint(
-                            UNUSED_IMPORTS,
-                            id,
-                            import.span,
-                            BuiltinLintDiag::RedundantImportVisibility {
-                                max_vis: max_vis.to_string(def_id, self.tcx),
-                                import_vis: import.vis.to_string(def_id, self.tcx),
-                                span: import.span,
-                            },
-                        );
-                    }
-                    return None;
+        let (ident, target, bindings, type_ns_only, import_id) = match import.kind {
+            ImportKind::Single { source, target, ref bindings, type_ns_only, id, .. } => {
+                (source, target, bindings, type_ns_only, id)
+            }
+            ImportKind::Glob { is_prelude, ref max_vis, id } => {
+                if import.module_path.len() <= 1 {
+                    // HACK(eddyb) `lint_if_path_starts_with_module` needs at least
+                    // 2 segments, so the `resolve_path` above won't trigger it.
+                    let mut full_path = import.module_path.clone();
+                    full_path.push(Segment::from_ident(Ident::dummy()));
+                    self.lint_if_path_starts_with_module(Some(finalize), &full_path, None);
                 }
-                _ => unreachable!(),
-            };
+
+                if let ModuleOrUniformRoot::Module(module) = module
+                    && module == import.parent_scope.module
+                {
+                    // Importing a module into itself is not allowed.
+                    return Some(UnresolvedImportError {
+                        span: import.span,
+                        label: Some(String::from("cannot glob-import a module into itself")),
+                        note: None,
+                        suggestion: None,
+                        candidates: None,
+                        segment: None,
+                        module: None,
+                    });
+                }
+                if !is_prelude
+                    && let Some(max_vis) = max_vis.get()
+                    && !max_vis.is_at_least(import.vis, self.tcx)
+                {
+                    let def_id = self.local_def_id(id);
+                    self.lint_buffer.buffer_lint(
+                        UNUSED_IMPORTS,
+                        id,
+                        import.span,
+                        BuiltinLintDiag::RedundantImportVisibility {
+                            max_vis: max_vis.to_string(def_id, self.tcx),
+                            import_vis: import.vis.to_string(def_id, self.tcx),
+                            span: import.span,
+                        },
+                    );
+                }
+                return None;
+            }
+            _ => unreachable!(),
+        };
 
         if self.privacy_errors.len() != privacy_errors_len {
             // Get the Res for the last element, so that we can point to alternative ways of
@@ -1095,17 +1088,17 @@ impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
                     ns,
                     &import.parent_scope,
                     Some(Finalize { report_private: false, ..finalize }),
-                    target_bindings[ns].get(),
+                    bindings[ns].get().binding(),
                     Some(import),
                 );
 
                 match binding {
                     Ok(binding) => {
                         // Consistency checks, analogous to `finalize_macro_resolutions`.
-                        let initial_res = source_bindings[ns].get().map(|initial_binding| {
+                        let initial_res = bindings[ns].get().binding().map(|binding| {
+                            let initial_binding = binding.import_source();
                             all_ns_err = false;
-                            if let Some(target_binding) = target_bindings[ns].get()
-                                && target.name == kw::Underscore
+                            if target.name == kw::Underscore
                                 && initial_binding.is_extern_crate()
                                 && !initial_binding.is_import()
                             {
@@ -1114,7 +1107,7 @@ impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
                                 } else {
                                     Used::Other
                                 };
-                                this.record_use(ident, target_binding, used);
+                                this.record_use(ident, binding, used);
                             }
                             initial_binding.res()
                         });
@@ -1126,7 +1119,7 @@ impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
                                 .span_delayed_bug(import.span, "some error happened for an import");
                             return;
                         }
-                        if let Ok(initial_res) = initial_res {
+                        if let Some(initial_res) = initial_res {
                             if res != initial_res {
                                 span_bug!(import.span, "inconsistent resolution for an import");
                             }
@@ -1269,7 +1262,7 @@ impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
         let mut any_successful_reexport = false;
         let mut crate_private_reexport = false;
         self.per_ns(|this, ns| {
-            let Ok(binding) = source_bindings[ns].get() else {
+            let Some(binding) = bindings[ns].get().binding().map(|b| b.import_source()) else {
                 return;
             };
 
@@ -1340,7 +1333,7 @@ impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
             let mut full_path = import.module_path.clone();
             full_path.push(Segment::from_ident(ident));
             self.per_ns(|this, ns| {
-                if let Ok(binding) = source_bindings[ns].get() {
+                if let Some(binding) = bindings[ns].get().binding().map(|b| b.import_source()) {
                     this.lint_if_path_starts_with_module(Some(finalize), &full_path, Some(binding));
                 }
             });
@@ -1350,7 +1343,7 @@ impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
         // this may resolve to either a value or a type, but for documentation
         // purposes it's good enough to just favor one over the other.
         self.per_ns(|this, ns| {
-            if let Ok(binding) = source_bindings[ns].get() {
+            if let Some(binding) = bindings[ns].get().binding().map(|b| b.import_source()) {
                 this.import_res_map.entry(import_id).or_default()[ns] = Some(binding.res());
             }
         });
@@ -1361,10 +1354,7 @@ impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
 
     pub(crate) fn check_for_redundant_imports(&mut self, import: Import<'ra>) -> bool {
         // This function is only called for single imports.
-        let ImportKind::Single {
-            source, target, ref source_bindings, ref target_bindings, id, ..
-        } = import.kind
-        else {
+        let ImportKind::Single { source, target, ref bindings, id, .. } = import.kind else {
             unreachable!()
         };
 
@@ -1391,7 +1381,8 @@ impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
         let mut is_redundant = true;
         let mut redundant_span = PerNS { value_ns: None, type_ns: None, macro_ns: None };
         self.per_ns(|this, ns| {
-            if is_redundant && let Ok(binding) = source_bindings[ns].get() {
+            let binding = bindings[ns].get().binding().map(|b| b.import_source());
+            if is_redundant && let Some(binding) = binding {
                 if binding.res() == Res::Err {
                     return;
                 }
@@ -1402,7 +1393,7 @@ impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
                     &import.parent_scope,
                     None,
                     false,
-                    target_bindings[ns].get(),
+                    bindings[ns].get().binding(),
                     None,
                 ) {
                     Ok(other_binding) => {

--- a/compiler/rustc_resolve/src/lib.rs
+++ b/compiler/rustc_resolve/src/lib.rs
@@ -891,6 +891,13 @@ impl<'ra> NameBindingData<'ra> {
         }
     }
 
+    fn import_source(&self) -> NameBinding<'ra> {
+        match self.kind {
+            NameBindingKind::Import { binding, .. } => binding,
+            _ => unreachable!(),
+        }
+    }
+
     fn is_ambiguity_recursive(&self) -> bool {
         self.ambiguity.is_some()
             || match self.kind {

